twilight-gateway/src/shard.rs

```diff
@@ -1219,70 +717,245 @@ impl<Q: Queue> Shard<Q> {
+                }
+                ready!(self.poll_flush_pending(cx))?;
             }
-        );
 
-        let fatal_code = CloseCode::AuthenticationFailed;
-        let fatal_status = ConnectionStatus::from_close_code(Some(fatal_code as u16));
+            let ratelimited = self
+                .ratelimiter
+                .as_mut()
+                .map_or(false, |ratelimiter| ratelimiter.poll_ready(cx).is_pending());
 
-        assert_eq!(
-            fatal_status,
-            ConnectionStatus::FatallyClosed {
-                close_code: fatal_code
+            if !ratelimited {
+                if let Some(Poll::Ready(canceled)) = self
+                    .identify_rx
+                    .as_mut()
+                    .map(|rx| Pin::new(rx).poll(cx).map(|r| r.is_err()))
+                {
+                    if canceled {
+                        self.identify_rx = Some(self.config.queue().enqueue(self.id.number()));
+                        continue;
+                    }
+                    tracing::debug!("sending identify");
+                    self.identify_rx = None;
+                    self.pending = Pending::text(
+                        json::to_string(&Identify::new(IdentifyInfo {
+                            compress: false,
+                            intents: self.config.intents(),
+                            large_threshold: self.config.large_threshold(),
+                            presence: self.config.presence().cloned(),
+                            properties: self
+                                .config
+                                .identify_properties()
+                                .cloned()
+                                .unwrap_or_else(default_identify_properties),
+                            shard: Some(self.id),
+                            token: self.config.token().to_owned(),
+                        }))
+                        .unwrap(),
+                        false,
+                    );
+                    ready!(self.poll_flush_pending(cx))?;
+                }
             }
-        );
 
-        let unknown_code = u16::MAX;
-        let non_fatal_status = ConnectionStatus::from_close_code(Some(unknown_code));
+            if !ratelimited && self.state.is_identified() {
+                if let Poll::Ready(command) = self.user_channel.command_rx.poll_recv(cx) {
+                    let command = command.expect("shard owns channel");
+                    tracing::debug!("sending command from user channel");
+                    self.pending = Pending::text(command, false);
+                    ready!(self.poll_flush_pending(cx))?;
+                }
+            }
 
-        assert_eq!(
-            non_fatal_status,
-            ConnectionStatus::Disconnected {
-                close_code: Some(unknown_code),
-                reconnect_attempts: 0
+            match ready!(Pin::new(self.connection.as_mut().expect("connected")).poll_next(cx)) {
```

Comment:
Either use `.unwrap()` or `.expect("connected")`, but don't mix both. And can we please de-duplicate the unwraps and assign to this unwrapped connection once at the start?